consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,241 @@
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode currentCheckpointId function parameters: %w", err)
+	}
+
+	currentCheckpointID, err := c.rootchain.Call(c.sender, helper.CheckpointManagerAddress, checkpointIDMethodEncoded)
+	if err != nil {
+		return 0, fmt.Errorf("failed to invoke currentCheckpointId function on the rootchain: %w", err)
+	}
+
+	checkpointID, err := strconv.ParseUint(currentCheckpointID, 0, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert current checkpoint id '%s' to number: %w",
+			currentCheckpointID, err)
+	}
+
+	return checkpointID, nil
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, isEndOfEpoch bool) error {
+	lastCheckpointBlockNumber, err := c.getCurrentCheckpointID()
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	initialBlockNumber := lastCheckpointBlockNumber + 1
+
+	var parentExtra *Extra
+
+	var parentHeader *types.Header
+
+	if initialBlockNumber < latestHeader.Number {
+		found := false
+		parentHeader, found = c.blockchain.GetHeaderByNumber(lastCheckpointBlockNumber)
+
+		if !found {
+			return fmt.Errorf("block %d was not found", lastCheckpointBlockNumber)
+		}
+
+		parentExtra, err = GetIbftExtra(parentHeader.ExtraData)
+		if err != nil {
+			return err
+		}
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNumber := initialBlockNumber; blockNumber < latestHeader.Number; blockNumber++ {
+		currentHeader, found := c.blockchain.GetHeaderByNumber(blockNumber)
+		if !found {
+			return fmt.Errorf("block %d was not found", blockNumber)
+		}
+
+		currentExtra, err := GetIbftExtra(currentHeader.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		parentEpochNumber := parentExtra.Checkpoint.EpochNumber
+		currentEpochNumber := currentExtra.Checkpoint.EpochNumber
+		parentHeader = currentHeader
+		parentExtra = currentExtra
+
+		// send pending checkpoints only for epoch ending blocks
+		if parentEpochNumber == currentEpochNumber {
+			continue
+		}
+
+		err = c.submitCheckpointInternal(nonce, txn, *parentHeader, *parentExtra, true)
+		if err != nil {
+			return err
+		}
+
+		nonce++
+	}
+
+	//we need to send checkpoint for the latest block
+	extra, err := GetIbftExtra(latestHeader.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	return c.submitCheckpointInternal(nonce, txn, latestHeader, *extra, isEndOfEpoch)
+}
+
+// submitCheckpointInternal encodes checkpoint data for the given block and
+// sends a transaction to the CheckpointManager rootchain contract
+func (c *checkpointManager) submitCheckpointInternal(nonce uint64, txn *ethgo.Transaction,
```

Comment:
A little bit confusing semantics, can this be `submitCheckpointTxn` or `submitCheckpointRootchain` or `sendToRootchain` `encodeAndSend`?

Reply:
Fixed [9db0e5e](https://github.com/0xPolygon/polygon-edge/pull/861/commits/9db0e5e4367c209d373969accd0e7e2c1f33d1fc)